contracts/staking/contracts/src/staking_pools/MixinCumulativeRewards.sol

```diff
@@ -0,0 +1,230 @@
+/*
+
+  Copyright 2019 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/LibFractions.sol";
+import "@0x/contracts-utils/contracts/src/LibSafeMath.sol";
+import "../immutable/MixinStorage.sol";
+import "../immutable/MixinConstants.sol";
+import "../stake/MixinStakeBalances.sol";
+import "./MixinStakingPoolRewardVault.sol";
+
+
+contract MixinCumulativeRewards is
+    IStakingEvents,
+    MixinConstants,
+    Ownable,
+    MixinStorage,
+    MixinZrxVault,
+    MixinStakingPoolRewardVault,
+    MixinScheduler,
+    MixinStakeStorage,
+    MixinStakeBalances
+{
+    using LibSafeMath for uint256;
+
+    /// @dev Initializes Cumulative Rewards for a given pool.
+    function _initializeCumulativeRewards(bytes32 poolId)
+        internal
+    {
+        uint256 currentEpoch = getCurrentEpoch();
+        _setCumulativeReward(poolId, currentEpoch, IStructs.Fraction({numerator: 0, denominator: MIN_TOKEN_VALUE}));
+        _setMostRecentCumulativeReward(poolId, currentEpoch);
+    }
+
+    /// @dev returns true iff Cumulative Rewards are set
+    function _isCumulativeRewardSet(IStructs.Fraction memory cumulativeReward)
+        internal
+        pure
+        returns (bool)
+    {
+        // we use the denominator as a proxy for whether the cumulative
+        // reward is set, as setting the cumulative reward always sets this
+        // field to at least 1.
+        return cumulativeReward.denominator != 0;
+    }
+
+    /// @dev Sets a cumulative reward for `poolId` at `epoch`.
+    /// @param poolId Unique Id of pool.
+    /// @param epoch of cumulative reward.
+    /// @param value of cumulative reward.
+    function _setCumulativeReward(
+        bytes32 poolId,
+        uint256 epoch,
+        IStructs.Fraction memory value
+    )
+        internal
+    {
+        cumulativeRewardsByPool[poolId][epoch] = value;
+    }
+
+    /// @dev Unsets the cumulative reward for `poolId` at `epoch`.
+    /// @param poolId Unique id of pool.
+    /// @param epoch of cumulative reward to unset.
+    function _unsetCumulativeReward(bytes32 poolId, uint256 epoch)
+        internal
+    {
+        assert(cumulativeRewardsByPoolReferenceCounter[poolId][epoch] == 0);
```

Comment:
If we adopt the encapsulation outlined below for reference counting, this assert can be removed.